pkg/virt-controller/watch/vm.go

```diff
@@ -924,6 +935,79 @@ func (c *VMController) stopVMI(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMac
 	return nil
 }
 
+func vmRevisionNamePrefix(vmName string) string {
+	return fmt.Sprintf("revision-start-vm-%s", vmName)
+}
+
+func getVMRevisionName(vmName string, generation int64) string {
+	return fmt.Sprintf("%s-%d", vmRevisionNamePrefix(vmName), generation)
+}
+
+func getVMRevisionData(vm *virtv1.VirtualMachine) (runtime.RawExtension, error) {
+	vmCpy := vm.DeepCopy()
+	vmCpy.Status = virtv1.VirtualMachineStatus{}
+	data, err := json.Marshal(vmCpy)
+	if err != nil {
+		return runtime.RawExtension{}, err
+	}
+	return runtime.RawExtension{Raw: data}, nil
+}
+
+func (c *VMController) deleteOlderVMRevision(vm *virtv1.VirtualMachine) (bool, error) {
+	labelSelector := fmt.Sprintf(virtv1.CreatedByLabel + "=" + string(vm.UID))
+	vmRevisions, err := c.clientset.AppsV1().ControllerRevisions(vm.Namespace).List(context.Background(), v1.ListOptions{LabelSelector: labelSelector})
```

Comment:
Although this is an efficient list with the labelselector we may want to watch/cache these to minimize calls to the k8s apiserver.  Look at vmsnapshot informer for example of indexing.